File: shell/ui/src/notify.ts
/**
 * Desktop notifications: what to post, and whether macOS will show it.
 *
 * The wording lives here rather than in Rust for the same reason the socket
 * types do: this is product copy, it changes with the design, and it is worth
 * being able to test on any machine. Rust posts what it is handed and reports
 * which button came back.
 *
 * Two rules shape everything below.
 *
 * The first is that a notification is a claim that something needs you. Only
 * the access request carries buttons, because it is the only event where the
 * answer is a decision rather than an acknowledgement - and it is the one the
 * marketing page promises you can settle "without opening the app".
 *
 * The second is that we never say permission is granted without having asked
 * macOS. [`NotifyState`] has a state for every honest answer, including the two
 * that are nobody's fault: a development build with no bundle, and a platform
 * this is not built for yet.
 */
import { useCallback, useEffect, useRef } from "react";
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { invoke } from "@tauri-apps/api/core";
import { listen } from "@tauri-apps/api/event";
import type { AccessEvent } from "./activity";
import { call } from "./daemon";
import { type Device, describeAgent } from "./devices";
import { EVENT } from "./events";
import type { App, Event, NotifyPrefs } from "./generated";

/** The Tauri event carrying a pressed button back. Matches notify.rs. */
const ACTION_EVENT = "kt://notification-action";

/**
 * What macOS will do with a notification from this app right now.
 *
 * `unbundled` is the state a developer sees: `pnpm tauri dev` runs a bare
 * binary, which has no bundle identifier for the notification framework to
 * attach to. `denied` cannot be undone from in here - macOS asks once and
 * remembers - so the surface points at System Settings instead of offering a
 * button that would do nothing.
 */
export interface NotifyState {
  state: "unsupported" | "unbundled" | "not_determined" | "denied" | "granted";
  /**
   * Whether banners are switched on, which is a separate setting from
   * permission. Granted with banners off means everything we post lands in
   * Notification Center unseen, and the surface has to say so rather than
   * showing a contented green tick.
   */
  alerts: boolean;
}

/** One notification, as Rust takes it. */
export interface Spec {
  title: string;
  body: string;
  subtitle?: string;
  /** Opaque to Rust; comes back on the action event. Always a device id here. */
  key?: string;
  actions?: Array<{ id: string; label: string }>;
  sound?: boolean;
}

export const APPROVE = "approve";
export const DENY = "deny";

/**
 * The banner for a device at the door, worded as the mockup draws it.
 *
 * The sentence is the title and the fingerprint line is the body, which is the
 * same split the in-app row uses: the sentence says what is happening, the mono
 * line is what you check the physical device against. Somebody comparing the
 * banner with the phone in their hand needs the second one.
 *
 * The app name is optional because it genuinely may not be known: devices are
 * trusted across the workspace and carry no app of their own, so the app comes
 * from the `requested` row in the access log, and a request can arrive before
 * that row is readable. "wants access" is true in both cases; inventing an app
 * name would not be.
 */
export function pairingSpec(device: Device, appName?: string): Spec {
  return {
    title: appName
      ? `${device.name} wants to open ${appName}`
      : `${device.name} wants access`,
    body: `fingerprint · ${device.fingerprint} · ${describeAgent(device.user_agent)}`,
    key: device.id,
    actions: [
      { id: APPROVE, label: "Approve" },
      { id: DENY, label: "Deny" },
    ],
    // The only notification this product makes a sound for. It is the only one
    // where somebody else is waiting on the answer.
    sound: true,
  };
}

/** "Priya's iPhone opened Trip Planner." Off by default; the mockup calls it chatty. */
export function openSpec(deviceName: string, appName: string): Spec {
  return {
    title: `${deviceName} opened ${appName}`,
    body: appName,
    subtitle: undefined,
  };
}

/**
 * Whether an event is worth a banner, given what the owner asked for.
 *
 * `deploys` and `offline` are stored and honoured and can never fire: nothing
 * writes a deploy (product.md 5.10) and nothing detects an app going offline.
 * They are switches over events that do not exist yet rather than switches that
 * are broken, and the surface says which is which.
 */
export function wanted(event: Event, prefs: NotifyPrefs): boolean {
  switch (event.event) {
    case "pairing_request":
      return prefs.requests;
    case "access":
      return prefs.opens && event.action === "opened";
    default:
      return false;
  }
}

/** The most recent app a device asked for, from the access log. */
export function requestedApp(
  log: AccessEvent[],
  deviceId: string,
): string | undefined {
  return log.find(
    (entry) => entry.device_id === deviceId && entry.action === "requested",
  )?.app_slug ?? undefined;
}

/** Ask Rust what macOS will do, without prompting anybody. */
export function useNotifyState() {
  return useQuery({
    queryKey: ["notify-state"],
    queryFn: () => invoke<NotifyState>("kt_notify_state"),
    // Someone can change this in System Settings while the window is open, and
    // the window would otherwise keep claiming notifications work.
    refetchInterval: 30_000,
    refetchOnWindowFocus: true,
    retry: false,
  });
}

/**
 * Show the system prompt.
 *
 * macOS asks once per install and answers from its own record afterwards, so
 * this is not a retry: a surface that offers it twice is offering a button that
 * does nothing the second time.
 */
export function useAskPermission() {
  const client = useQueryClient();
  return useMutation({
    mutationFn: () => invoke<NotifyState>("kt_notify_request"),
    onSuccess: (state) => client.setQueryData(["notify-state"], state),
  });
}

export function useNotifyPrefs() {
  return useQuery({
    queryKey: ["notify-prefs"],
    queryFn: () => call<NotifyPrefs>("notify.prefs"),
    retry: false,
  });
}

export function useSetNotifyPrefs() {
  const client = useQueryClient();
  return useMutation({
    mutationFn: (patch: Partial<NotifyPrefs>) =>
      call<NotifyPrefs>("notify.set_prefs", patch),
    // The daemon answers with all four, so the cache takes its word rather
    // than the patch we sent.
    onSuccess: (prefs) => client.setQueryData(["notify-prefs"], prefs),
  });
}

/** Post one notification. Silently does nothing where they are unsupported. */
export function show(spec: Spec): void {
  void invoke("kt_notify", { spec }).catch(() => {
    // No IPC - the UI also runs in a plain browser and under vitest. A missing
    // banner is not worth taking the window down for.
  });
}

/**
 * Post banners for daemon events, and act on the buttons.
 *
 * Mounted once, beside [`useDaemonEvents`], and for the same reason: an event
 * that arrives while the Notifications surface is closed still has to reach the
 * desktop. The window being hidden is the normal case here - that is what the
 * whole feature is for.
 *
 * Approving from a banner goes back over the socket exactly as the in-app
 * button does. There is no second path into that decision, which is what keeps
 * the access log honest about how a device got in.
 */
export function useDesktopNotifications(prefs?: NotifyPrefs): void {
  const client = useQueryClient();

  // Read through a ref so the subscription is not torn down and rebuilt every
  // time a switch moves; a re-subscribe races with the events it exists to
  // catch.
  const current = useRef(prefs);
  current.current = prefs;

  const post = useCallback(async (event: Event) => {
    const prefs = current.current;
    if (!prefs || !wanted(event, prefs)) return;

    if (event.event === "pairing_request") {
      const devices = await call<Device[]>("device.list");
      const device = devices.find((candidate) => candidate.id === event.device_id);
      // Gone already: approved from another window, or denied between the
      // event and this call. Nothing to ask about.
      if (!device || device.status !== "pending") return;

      const log = await call<AccessEvent[]>("log.query", { limit: 50 });
      const slug = requestedApp(log, device.id);
      const apps = slug ? await call<App[]>("app.list") : [];
      show(pairingSpec(device, apps.find((app) => app.slug === slug)?.name ?? slug));
      return;
    }

    if (event.event === "access" && event.app_slug) {
      const apps = await call<App[]>("app.list");
      const app = apps.find((candidate) => candidate.slug === event.app_slug);
      if (app) show(openSpec("Someone", app.name));
    }
  }, []);

  useEffect(() => {
    let stops: Array<() => void> = [];
    let gone = false;

    try {
      void Promise.all([
        listen<Event>(EVENT, ({ payload }) => void post(payload)),
        listen<{ key?: string; action: string }>(ACTION_EVENT, ({ payload }) => {
          if (!payload.key) return;
          const method =
            payload.action === APPROVE
              ? "device.approve"
              : payload.action === DENY
                ? "device.deny"
                : null;
          // Dismissed, timed out, or the body was clicked. All of those leave
          // the request standing, which is the safe direction: a device is let
          // in only by somebody saying so.
          if (!method) return;

          void call(method, { id: payload.key })
            .catch(() => {
              // Answered elsewhere in the meantime. The list below is right
              // either way, so the refetch covers it.
            })
            .finally(() => {
              void client.invalidateQueries({ queryKey: ["devices"] });
              void client.invalidateQueries({ queryKey: ["log"] });
            });
        }),
      ])
        .then((listeners) => {
          if (gone) listeners.forEach((stop) => stop());
          else stops = listeners;
        })
        .catch(() => {
          // No IPC. Same as everywhere else: the window still works.
        });
    } catch {
      // Same.
    }

    return () => {
      gone = true;
      stops.forEach((stop) => stop());
    };
  }, [client, post]);
}

File: shell/src-tauri/src/notify.rs
//! Desktop notifications, and the permission macOS gates them behind.
//!
//! The shell owns this rather than the daemon for the same reason it owns the
//! tray: it is the half with a user session to draw on. The daemon has no
//! desktop, and a headless install has nobody to notify.
//!
//! It stays a proxy all the same (docs/architecture.md section 13). Rust posts
//! whatever wording the window hands it and reports which button came back; it
//! does not decide what an access request is, or what approving one means. That
//! knowledge stays in one place, on the socket, where the CLI and an agent can
//! reach it too.
//!
//! # Why not `tauri-plugin-notification`
//!
//! It was tried first. On desktop it sets a title, body, icon and sound and
//! nothing else, so the Approve and Deny buttons the design puts on an access
//! request cannot exist; and its `permission_state` is a stub that returns
//! `Granted` without asking macOS, so a window built on it would draw a
//! settled permission it had never checked. Both matter here, so this talks to
//! `UNUserNotificationCenter` instead - the framework Apple supports, and the
//! only one with an authorisation prompt.
//!
//! # What macOS requires
//!
//! A bundled, signed app. `pnpm tauri dev` runs a bare binary with no bundle
//! identifier, and there the whole surface reports [`UNBUNDLED`] rather than
//! failing: notifications are the one part of this product that cannot be
//! developed against the dev server, and saying so is better than a permission
//! prompt that never appears and an error nobody can act on.

use serde::Deserialize;

/// The Tauri event carrying a pressed button back to the window.
pub const ACTION_EVENT: &str = "kt://notification-action";

/// Not macOS. Notifications are not built for this platform yet.
#[cfg_attr(target_os = "macos", allow(dead_code))]
const UNSUPPORTED: &str = "unsupported";
/// macOS, but this process has no bundle identifier - a `tauri dev` binary.
const UNBUNDLED: &str = "unbundled";
/// Nobody has been asked yet. The window offers a button that asks.
const NOT_DETERMINED: &str = "not_determined";
/// The owner said no, or Focus/Do Not Disturb settings amount to the same.
/// Only System Settings can undo it; asking again does nothing, by design.
const DENIED: &str = "denied";
const GRANTED: &str = "granted";

/// How long an actionable notification waits for an answer.
///
/// Not optional: `UNUserNotificationCenter` never reports a response for a
/// notification the owner cleared from Notification Center, so a request with
/// no timeout can leave a thread parked for the life of the process. Five
/// minutes is long enough for somebody to walk back to their desk, and when it
/// lapses the banner is withdrawn and the request is still waiting in the
/// window - which is where a decision this old belongs anyway.
#[cfg(target_os = "macos")]
const ANSWER_WITHIN: std::time::Duration = std::time::Duration::from_secs(300);

/// One notification, as the window describes it.
///
/// `key` is opaque here and comes back untouched on the action event: the
/// window sends a device id, and Rust never learns that is what it is.
#[derive(Debug, Deserialize)]
pub struct Spec {
    pub title: String,
    pub body: String,
    #[serde(default)]
    pub subtitle: Option<String>,
    #[serde(default)]
    pub key: Option<String>,
    /// Buttons, in the order they should appear. Empty means an informational
    /// banner with nothing to press.
    #[serde(default)]
    pub actions: Vec<Button>,
    #[serde(default)]
    pub sound: bool,
}

#[derive(Debug, Deserialize)]
pub struct Button {
    /// Sent back verbatim on the action event.
    pub id: String,
    pub label: String,
}

#[cfg(target_os = "macos")]
mod imp {
    use super::{Spec, ACTION_EVENT, ANSWER_WITHIN, DENIED, GRANTED, NOT_DETERMINED, UNBUNDLED};
    use mac_usernotifications::{
        block_on, blocking, check_bundle, Action, AuthorizationStatus, Notification,
        NotificationSettingStatus,
    };
    use tauri::Emitter;

    /// What macOS will currently do with a notification from this app.
    ///
    /// `alerts` is reported separately from the authorisation because the two
    /// really do come apart: an owner can grant permission and still have
    /// banners switched off, in which case everything we post goes straight to
    /// Notification Center unseen. A window that said "on" there would be
    /// telling somebody their access requests will interrupt them when they
    /// will not.
    pub fn state() -> serde_json::Value {
        if check_bundle().is_err() {
            return serde_json::json!({ "state": UNBUNDLED, "alerts": false });
        }

        match blocking::get_notification_settings() {
            Ok(settings) => {
                let state = match settings.authorization_status {
                    AuthorizationStatus::Authorized
                    | AuthorizationStatus::Provisional
                    | AuthorizationStatus::Ephemeral => GRANTED,
                    AuthorizationStatus::Denied => DENIED,
                    // An unknown status is treated as "not asked". The prompt
                    // is harmless if it turns out we have permission already -
                    // macOS shows it once and answers from its own record
                    // afterwards - and this way a future status we do not know
                    // about cannot silently turn notifications off.
                    AuthorizationStatus::NotDetermined | AuthorizationStatus::Unknown => {
                        NOT_DETERMINED
                    }
                };
                serde_json::json!({
                    "state": state,
                    "alerts": settings.alert_enabled == NotificationSettingStatus::Enabled,
                })
            }
            Err(e) => {
                tracing::warn!(error = %e, "could not read the notification settings");
                serde_json::json!({ "state": NOT_DETERMINED, "alerts": false })
            }
        }
    }

    /// Show the system prompt, then report where that left us.
    ///
    /// macOS only ever asks once per install; after that this returns the
    /// stored answer without showing anything. That is why the window must not
    /// offer it as a retry - a second press would look broken.
    pub fn request() -> serde_json::Value {
        if check_bundle().is_err() {
            return serde_json::json!({ "state": UNBUNDLED, "alerts": false });
        }

        match blocking::request_auth() {
            Ok(granted) => tracing::info!(granted, "notification permission answered"),
            Err(e) => tracing::warn!(error = %e, "could not ask for notification permission"),
        }

        // Deliberately re-read rather than trusting the boolean: the prompt
        // answers authorisation only, and the window also needs to know whether
        // banners are switched on.
        state()
    }

    /// Post one notification and, if it has buttons, wait for the answer.
    ///
    /// Off the calling thread, because waiting is the normal case: the future
    /// resolves when somebody presses a button, which may be minutes away. The
    /// delegate callback arrives on the main run loop, which Tauri is already
    /// pumping.
    pub fn show(app: tauri::AppHandle, spec: Spec) {
        if check_bundle().is_err() {
            tracing::debug!("no bundle identifier; skipping the notification");
            return;
        }

        std::thread::spawn(move || {
            let mut notification = Notification::new().title(&spec.title).message(&spec.body);

            if let Some(subtitle) = &spec.subtitle {
                notification = notification.subtitle(subtitle);
            }
            if spec.sound {
                notification = notification.default_sound();
            }

            let actionable = !spec.actions.is_empty();
            for button in &spec.actions {
                notification = notification.action(Action::button(&button.id, &button.label));
            }
            if actionable {
                notification = notification.timeout(ANSWER_WITHIN);
            }

            let handle = match notification.send_blocking() {
                Ok(handle) => handle,
                Err(e) => {
                    // Includes the case where the owner has denied permission:
                    // there is nothing to recover here, and the window already
                    // shows the same request in a list.
                    tracing::warn!(error = %e, "could not post a notification");
                    return;
                }
            };

            // Nothing to wait for, and nothing the window could do with a
            // dismissal of a banner it only meant as information.
            if !actionable {
                return;
            }

            match block_on(handle.response()) {
                Ok(response) => {
                    // The body was clicked rather than a button. Window
                    // management is the shell's own job, so this is the one
                    // response it acts on itself - and it still forwards the
                    // event, because the window may want to show the request.
                    if response.is_default_action() {
                        crate::show_window(&app);
                    }

                    let action = if response.is_default_action() {
                        "default".to_string()
                    } else if response.is_timed_out() {
                        "timed_out".to_string()
                    } else if response.is_dismiss_action() {
                        "dismissed".to_string()
                    } else {
                        response.action_identifier.clone()
                    };

                    if let Err(e) = app.emit(
                        ACTION_EVENT,
                        serde_json::json!({ "key": spec.key, "action": action }),
                    ) {
                        tracing::warn!(error = %e, "could not deliver a notification action");
                    }
                }
                Err(e) => tracing::warn!(error = %e, "no answer from the notification"),
            }
        });
    }
}

#[cfg(not(target_os = "macos"))]
mod imp {
    use super::{Spec, UNSUPPORTED};

    pub fn state() -> serde_json::Value {
        serde_json::json!({ "state": UNSUPPORTED, "alerts": false })
    }

    pub fn request() -> serde_json::Value {
        state()
    }

    /// Deliberately silent. Windows and Linux get their own path when this
    /// product ships for them; until then the window reads [`UNSUPPORTED`] and
    /// says so, rather than offering a switch that does nothing.
    pub fn show(_app: tauri::AppHandle, _spec: Spec) {}
}

/// Whether macOS will show a notification from this app, without asking.
#[tauri::command]
pub fn kt_notify_state() -> serde_json::Value {
    imp::state()
}

/// Ask macOS for permission, showing the system prompt if it has never asked.
#[tauri::command]
pub fn kt_notify_request() -> serde_json::Value {
    imp::request()
}

/// Post a notification. Answers to its buttons arrive on [`ACTION_EVENT`].
#[tauri::command]
pub fn kt_notify(app: tauri::AppHandle, spec: Spec) {
    imp::show(app, spec);
}

/// Kept honest by the fact that both halves of `imp` are compiled from the same
/// signatures; the useful assertions about wording and timing live in the UI's
/// own tests, where they can run on every platform.
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_spec_needs_only_a_title_and_a_body() {
        let spec: Spec = serde_json::from_value(serde_json::json!({
            "title": "Priya's iPad wants access",
            "body": "Trip Planner",
        }))
        .expect("deserialises");

        assert!(
            spec.actions.is_empty(),
            "an informational banner by default"
        );
        assert!(!spec.sound);
        assert_eq!(spec.key, None);
    }

    #[test]
    fn buttons_keep_the_order_the_window_sent() {
        let spec: Spec = serde_json::from_value(serde_json::json!({
            "title": "t",
            "body": "b",
            "key": "device-1",
            "actions": [
                { "id": "approve", "label": "Approve" },
                { "id": "deny", "label": "Deny" },
            ],
        }))
        .expect("deserialises");

        let ids: Vec<_> = spec.actions.iter().map(|a| a.id.as_str()).collect();
        assert_eq!(ids, ["approve", "deny"]);
        assert_eq!(spec.key.as_deref(), Some("device-1"));
    }

    #[test]
    fn the_state_is_always_a_string_the_window_can_switch_on() {
        let state = kt_notify_state();
        let named = state["state"].as_str().expect("a state string");

        assert!(
            [UNSUPPORTED, UNBUNDLED, NOT_DETERMINED, DENIED, GRANTED].contains(&named),
            "unexpected state {named:?}",
        );
    }
}
